Surface EIP-712 validation errors to developers

**What problem are you trying to solve?**
As a developer working on implementing EIP-712, it's very difficult when you don't know how you're mismatching the spec. It looks like there're some handy validation errors at this point in the code: https://github.com/MetaMask/metamask-extension/blob/934b10307c97793dd2cf5ac10c3b4d96fb8cbfcc/app/scripts/lib/typed-message-manager.js#L153 however they're not surfaced at all.

**Describe the solution you'd like**
It would be great to append those to the error that's thrown. For instance, I currently get:
```
inpage.js:1 MetaMask - RPC Error: AssertionError: Data must conform to EIP-712 schema. See https://git.io/fNtcx.
    at t.value (chrome-extension://dilihgaobegidgflofnafcidjkegelho/background.js:1:137305)
    ...
```
It'd be very helpful to get
```
inpage.js:1 MetaMask - RPC Error: AssertionError: Data must conform to EIP-712 schema. See https://git.io/fNtcx. You had 2 errors: "instance.types.message is not of a type(s) array", "instance.types.time is not of a type(s) array"
    at t.value (chrome-extension://dilihgaobegidgflofnafcidjkegelho/background.js:1:137305)
    ...
```

At the very least, it'd be great to throw in a console.error with the validation object so that we could inspect it in background.js.

--- a/app/scripts/lib/typed-message-manager.js
+++ b/app/scripts/lib/typed-message-manager.js
@@ -1,7 +1,7 @@
 import EventEmitter from 'events';
 import { strict as assert } from 'assert';
 import { ObservableStore } from '@metamask/obs-store';
-import { ethErrors } from 'eth-rpc-errors';
+import { errorCodes, EthereumProviderError, ethErrors } from 'eth-rpc-errors';
 import { typedSignatureHash, TYPED_MESSAGE_SCHEMA } from 'eth-sig-util';
 import log from 'loglevel';
 import jsonschema from 'jsonschema';
@@ -192,11 +192,18 @@ export default class TypedMessageManager extends EventEmitter {
           data.primaryType in data.types,
           `Primary type of "${data.primaryType}" has no type definition.`,
         );
-        assert.equal(
-          validation.errors.length,
-          0,
-          'Signing data must conform to EIP-712 schema. See https://git.io/fNtcx.',
-        );
+        if (validation.errors.length !== 0) {
+          throw new EthereumProviderError(
+            4009,
+            'Signing data must conform to EIP-712 schema. See https://git.io/fNtcx.',
+            validation.errors.map((v) => v.message.toString()),
+          );
+        }
+        // assert.equal(
+        //   validation.errors.length,
+        //   0,
+        //   `Signing data must conform to EIP-712 schema. See https://git.io/fNtcx. \n Errors: ${validation.errors.join('\n')}`,
+        // );
         let { chainId } = data.domain;
         if (chainId) {
           const activeChainId = parseInt(this._getCurrentChainId(), 16);

--- a/app/scripts/lib/typed-message-manager.js
+++ b/app/scripts/lib/typed-message-manager.js
@@ -1,7 +1,7 @@
 import EventEmitter from 'events';
 import { strict as assert } from 'assert';
 import { ObservableStore } from '@metamask/obs-store';
-import { errorCodes, EthereumProviderError, ethErrors } from 'eth-rpc-errors';
+import { EthereumRpcError, ethErrors } from 'eth-rpc-errors';
 import { typedSignatureHash, TYPED_MESSAGE_SCHEMA } from 'eth-sig-util';
 import log from 'loglevel';
 import jsonschema from 'jsonschema';
@@ -193,17 +193,12 @@ export default class TypedMessageManager extends EventEmitter {
           `Primary type of "${data.primaryType}" has no type definition.`,
         );
         if (validation.errors.length !== 0) {
-          throw new EthereumProviderError(
-            4009,
+          throw new EthereumRpcError(
+            -32602,
             'Signing data must conform to EIP-712 schema. See https://git.io/fNtcx.',
             validation.errors.map((v) => v.message.toString()),
           );
         }
-        // assert.equal(
-        //   validation.errors.length,
-        //   0,
-        //   `Signing data must conform to EIP-712 schema. See https://git.io/fNtcx. \n Errors: ${validation.errors.join('\n')}`,
-        // );
         let { chainId } = data.domain;
         if (chainId) {
           const activeChainId = parseInt(this._getCurrentChainId(), 16);

--- a/app/scripts/lib/typed-message-manager.js
+++ b/app/scripts/lib/typed-message-manager.js
@@ -194,7 +194,7 @@ export default class TypedMessageManager extends EventEmitter {
         );
         if (validation.errors.length !== 0) {
           throw new EthereumRpcError(
-            -32602,
+            ethErrors.rpc.invalidParams,
             'Signing data must conform to EIP-712 schema. See https://git.io/fNtcx.',
             validation.errors.map((v) => v.message.toString()),
           );

